source/crud/crud.rst

```diff
@@ -1927,8 +1928,63 @@ The ``update`` family of operations require that the update document parameter M
 
 
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+Aggregation Pipelines with Write Stages
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+This section discusses special considerations for aggregation pipelines that
+contain write stages (e.g. ``$out``, ``$merge``).
+
+
+Returning a cursor on the output collection
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+As noted in the documentation for the ``aggregate`` helper earlier in this
+document, ``$out`` and ``$merge`` are special pipeline stages that cause no
+results to be returned from the server. As such, drivers MAY setup a cursor to
+be executed upon iteration against the output collection and return that instead
+of an iterable that would otherwise have no results.
+
+
+Read preferences and server selection
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+Historically, only primaries could execute an aggregation pipeline with ``$out``
+or ``$merge``. As of ``featureCompatibilityVersion`` 4.4, secondaries can now
+execute such pipelines; however, since drivers do not track
+``featureCompatibilityVersion``, executing an aggregation pipeline with ``$out``
+or ``$merge`` on a secondary will require a 5.0+ server (i.e. wire version >=
+13).
+
+Drivers MUST guarantee the following:
+
+- If an explicit (i.e. per-operation) read preference is specified for an
+  aggregation with a write stage, drivers MUST attempt to use it. If that would
+  result in a pre-5.0, secondary server being selected, drivers MUST instead
+  select a server using a primary read preference.
+- If no explicit read preference is specified but a default read preference is
+  available to inherit (e.g. from the Collection), drivers MUST attempt to use
+  it. If that would result in a pre-5.0, secondary server being selected,
+  drivers MUST instead select a server using a primary read preference.
+
+Since the server version may not be known prior to server selection, this
+specification is not prescriptive about how drivers should select a server for
+an ``aggregate`` command including a write stage. For example, drivers MAY use a
+custom server selector to consider server/wire version when matching a read
+preference and fall back to selecting a primary. Alternatively, drivers MAY
+invoke server selection an additional time if necessary.
+
+Drivers MUST follow the rules for
+`Passing read preference to mongos and load balancers <../server-selection/server-selection.rst#passing-read-preference-to-mongos-and-load-balancers>`_.
+If an explicit or inherited read preference was used and the selected server is
+a pre-5.0 mongos or a load balancer backing a pre-5.0 mongos, drivers MUST
+behave as if a primary read preference was used. This is especially important if
+drivers utilize a custom server selector, since they may need to infer whether a
+primary read preference was used as a fall back.
```

Comment:
The goal of this paragraph is just the remind drivers that the effective read preference may need to be passed along to mongos or a load balancer. If the user-specified read preference ultimately wasn't used (e.g. it selected a pre-5.0 server), we should assume that a primary read preference is needed as a fall back. The [upgrade process for sharded clusters](https://docs.mongodb.com/manual/tutorial/upgrade-revision/#upgrade-sharded-clusters) sees mongos processes upgraded _after_ shards. Given that, I think the only safe assumption is that if mongos is 5.0+, then we can assume the servers behind it are also 5.0+ (but not vice versa).

I realize LBs were introduced with MongoDB 5.0, but I saw no harm in being pedantic about a "load balancer backing a pre-5.0 mongos".

Reply:
I am having a bit of trouble following this paragraph (the proposed change in the diff). Is this referring to the case of $merge/$out being redirected to primary on pre-5.0 and, subsequently, pointing out that since the query is being sent to primary the read preference that is sent along with the query must therefore also be primary?

Reply:
> Is this referring to the case of $merge/$out being redirected to primary on pre-5.0 and, subsequently, pointing out that since the query is being sent to primary the read preference that is sent along with the query must therefore also be primary?

Going back to the scope for this feature, we are cautiously assuming that a read preference can only be used for `$out`/`$merge` if the server itself is 5.0+. We aren't differentiating between the kind of read preference being used at all. If the driver ultimately selects a pre-5.0 mongos (either directly or through the LB), we will not pass a read preference at all. Erring on the side of caution, we're going to assume that a pre-5.0 mongos is fronting pre-5.0 mongod servers.

But if a 5.0+ mongos is selected, we can assume it's safe to pass along _any_ read preference. In this case, that will either be what was explicitly provided to the operation or inherited through some driver object (e.g. collection). That correlates with the two bullets in "Drivers MUST guarantee the following" a few paragraphs up.

This approach should not require any special knowledge of what happens _within_ server selection, be it a single attempt with custom logic or two separate attempts. To keep things simple, we need only consider the wire version of the server that was ultimately selected.

I suppose for drivers relying on two separate selection attempts, they can be more efficient. For example, if `mode=secondary` returns a pre-5.0 mongos, a driver could still use that mongos and just omit passing the read preference onwards. A second selection attempt with `mode=primary` would be a waste of time, since we'd likely just get another pre-5.0 mongos back (ignoring the possibility that there are mixed mongos versions available).

----

Going back to what you wrote in your other top-level PR comment:

> I looked through our (Ruby's) present server selection implementation and it seems to me that it would be relatively straightforward to add a condition along the lines of "if wire version is less than X use primary". However, our present implementation performs server selection twice if $out/$merge ends up being directed to a primary (hence could be subject to twice the timeout). I believe it would be more work to reduce this to a single round since we would need to add a mechanism for returning the primary read preference back from the server selector so that we could send the correct one to mongos/load balancers as mentioned.

I think I might have already answered this above, but there should be no need to change the server selector to return a read preference in addition to the selected server. If the first attempt comes back with a 5.0+ mongos, that's great and you can pass along the original read preference you provided the server selector (if any). If a pre-5.0 mongos is selected, you can assume no read preference should be pass on. In either case, I don't think you'd need to make a second selection attempt.

A second selection should only come up when dealing with a replica set topology where the first selection attempt returns a pre-5.0 _secondary_ server. If the user is trying to use `mode=nearest` and that happens to select a pre-5.0 primary, there's no reason to repeat server selection with `mode=primary` since we'd ultimately get the same result. That said, if `mode=nearest` happened to select a pre-5.0 secondary, we would need to repeat server selection to get a primary (or use a single attempt with custom logic).

Reply:
I reviewed our code again and it should only redirect in replica set topology (standalones and mongoses are always accepted on the first try). So I think we are in alignment here, I will resolve this thread.